### PR TITLE
225 - 227. Adding Validation in User Input

### PR DESCRIPTION
**Pull request by MapleSyyrup**
- Add validator per form field

image

### CODE REVIEW

**Comment by MapleSyyrup** [comment]
image

File: lib/screens/edit_product_screen.dart [hunk 1]
```diff
@@ -40,11 +40,23 @@ class _EditProductScreenState extends State<EditProductScreen> {
 
   void _updateImageUrl() {
     if (!_imageUrlFocusNode.hasFocus) {
+      if ((!_imageUrlController.text.startsWith('http') && !_imageUrlController.text.startsWith('https')) ||
+          (!_imageUrlController.text.endsWith('.png') &&
+              !_imageUrlController.text.endsWith('jpg') &&
+              !_imageUrlController.text.endsWith('.jpeg'))) {
+        return;
+      }
       setState(() {});
```
Comment (Nei1eveN) [inline lib/screens/edit_product_screen.dart:49]: remove this

File: lib/screens/edit_product_screen.dart [hunk 2]
```diff
@@ -86,6 +98,49 @@ class _EditProductScreenState extends State<EditProductScreen> {
...
+      return 'Please enter a number greater than zero.';
+    }
+    return null;
+  }
+
+  String validatorTitle(String value) {
+    if (value.isEmpty) {
+      return 'Please provide a value.';
+    }
+    return null;
```
Comment (Nei1eveN) [inline lib/screens/edit_product_screen.dart]: can you try if this is possible:

`return value.isEmpty ? 'Please provide a value' : null;`

File: lib/screens/edit_product_screen.dart [hunk 3]
```diff
@@ -86,6 +98,49 @@ class _EditProductScreenState extends State<EditProductScreen> {
...
+    if (value.isEmpty) {
+      return 'Please enter an image URL';
+    }
+    if (value.startsWith('http') && !value.startsWith('https')) {
+      return 'Please enter a valid URL.';
+    }
+    if (!value.endsWith('.png') && !value.endsWith('jpg') && !value.endsWith('.jpeg')) {
+      return 'Please enter a valid image URL.';
+    }
+    return null;
```
Comment (Nei1eveN) [inline lib/screens/edit_product_screen.dart:111]: try:

`return value.isEmpty ? 'Please enter an image URL' : !value.startsWith('https') ? 'Please enter a valid URL.' : !value.endsWith('.png') || !value.endsWith('jpg') || !value.endsWith('.jpeg') ? 'Please enter a valid image URL.' : null`

File: lib/screens/edit_product_screen.dart [hunk 4]
```diff
@@ -86,6 +98,49 @@ class _EditProductScreenState extends State<EditProductScreen> {
...
+  }
+
+  String validatorDescription(String value) {
+    if (value.isEmpty) {
+      return 'Please enter a description';
+    }
+    if (value.length < 10 != null) {
+      return 'Should be at least 10 characters long.';
+    }
+    return null;
```
Comment (Nei1eveN) [inline lib/screens/edit_product_screen.dart:121]: try:

`return value.isEmpty ? 'Please enter a description' : value.length < 10 != null ? 'Should be at least 10 characters long.' : null;`

File: lib/screens/edit_product_screen.dart [hunk 5]
```diff
@@ -86,6 +98,49 @@ class _EditProductScreenState extends State<EditProductScreen> {
...
+    if (value.isEmpty) {
+      return 'Please enter a price.';
+    }
+    if (double.tryParse(value) == null) {
+      return 'Please enter a valid number';
+    }
+    if (double.parse(value) <= 0) {
+      return 'Please enter a number greater than zero.';
+    }
+    return null;
```
Comment (Nei1eveN) [inline lib/screens/edit_product_screen.dart:134]: try:

`return value.isEmpty ? 'Please enter a price.' : double.tryParse(value) == null ? 'Please enter a valid number' : double.parse(value) <= 0 ? 'Please enter a number greater than zero.' : null;`

**Review by Nei1eveN — CHANGES_REQUESTED** [review]
see my comments

File: lib/screens/edit_product_screen.dart (hunk 3 above)
Comment (MapleSyyrup) [reply lib/screens/edit_product_screen.dart:111]: image

I tried implementing it but even if the URL starts with https and ends with jpg, it still shows that there is an error in the URL. Please see below.

image

File: lib/screens/edit_product_screen.dart (hunk 4 above)
Comment (MapleSyyrup) [reply lib/screens/edit_product_screen.dart:121]: image

Even if the characters are more than 10, it still shows the validation.

File: lib/screens/edit_product_screen.dart (hunk 3 above)
Comment (Nei1eveN) [reply lib/screens/edit_product_screen.dart:111]: @MapleSyyrup can you send here the value of image url you are trying to display?

File: lib/screens/edit_product_screen.dart (hunk 5 above)
Comment (MapleSyyrup) [reply lib/screens/edit_product_screen.dart:134]: it worked.

File: lib/screens/edit_product_screen.dart (hunk 1 above)
Comment (MapleSyyrup) [reply lib/screens/edit_product_screen.dart:49]: The setState cannot be removed. If the setState is removed, the preview of imageUrl wont be loaded unless the save button or the check button in the keyboard is pressed. 
setState is used to load the image when the user clicks other TextFormField aside from the imageUrl without the need to press the save button or check button.

image

File: lib/screens/edit_product_screen.dart (hunk 2 above)
Comment (MapleSyyrup) [reply lib/screens/edit_product_screen.dart]: it worked.

File: lib/screens/edit_product_screen.dart (hunk 4 above)
Comment (Nei1eveN) [reply lib/screens/edit_product_screen.dart:121]: I see. Then, we will not change anything here

File: lib/screens/edit_product_screen.dart (hunk 3 above)
Comment (MapleSyyrup) [reply lib/screens/edit_product_screen.dart:111]: https://upload.wikimedia.org/wikipedia/commons/thumb/1/14/Cast-Iron-Pan.jpg/1024px-Cast-Iron-Pan.jpg

Comment (Nei1eveN) [reply lib/screens/edit_product_screen.dart:111]: I see. Then just go with the original implementation for now.

**Review by Nei1eveN — APPROVED** [review]

**Pull request merged**